metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java

```diff
@@ -0,0 +1,262 @@
+        resetSnapshotCounters();
+    }
+
+    @Override
+    public String name() {
+        return "SnapshotGenerator";
+    }
+
+    void resetSnapshotCounters() {
+        this.bytesSinceLastSnapshot = 0L;
+        this.lastSnapshotTimeNs = time.nanoseconds();
+    }
+
+    @Override
+    public void publishSnapshot(
+        MetadataDelta delta,
+        MetadataImage newImage,
+        SnapshotManifest manifest
+    ) {
+        // Reset the snapshot counters because we just read a snapshot.
+        resetSnapshotCounters();
+    }
+
+    @Override
+    public void publishPreVersionChangeImage(
+        MetadataDelta delta,
+        MetadataImage preVersionChangeImage,
+        PreVersionChangeManifest manifest
+    ) {
+        scheduleEmit("the metadata version changed", preVersionChangeImage);
+    }
+
+    @Override
+    public void publishLogDelta(
+        MetadataDelta delta,
+        MetadataImage newImage,
+        LogDeltaManifest manifest
+    ) {
+        bytesSinceLastSnapshot += manifest.numBytes();
+        if (bytesSinceLastSnapshot >= minBytesSinceLastSnapshot) {
+            if (eventQueue.isEmpty()) {
+                scheduleEmit("we have replayed at least " + minBytesSinceLastSnapshot +
+                    " bytes", newImage);
+            }
+        } else if (minTimeSinceLastSnapshotNs != 0 &&
+                (time.nanoseconds() - lastSnapshotTimeNs >= minTimeSinceLastSnapshotNs)) {
+            if (eventQueue.isEmpty()) {
+                scheduleEmit("we have waited at least " +
+                    TimeUnit.NANOSECONDS.toMinutes(minTimeSinceLastSnapshotNs) + " minute(s)", newImage);
+            }
+        }
+    }
+
+    void scheduleEmit(
+        String reason,
+        MetadataImage image
+    ) {
+        resetSnapshotCounters();
+        eventQueue.append(() -> {
+            String currentDisabledReason = disabledReason;
+            if (currentDisabledReason != null) {
+                log.error("Not emitting {} despite the fact that {} because snapshots are " +
+                    "disabled; {}", image.provenance().snapshotName(), reason,
+                        currentDisabledReason);
+            } else {
+                log.info("Creating new KRaft snapshot file {} because {}.",
+                        image.provenance().snapshotName(), reason);
+                try {
+                    emitter.emit(image);
+                } catch (Throwable e) {
+                    faultHandler.handleFault("KRaft snapshot file generation error", e);
+                }
+            }
+        });
+    }
+
+    public void beginShutdown() {
+        this.disabledReason = "we are shutting down";
+        eventQueue.beginShutdown("beginShutdown");
+    }
+
+    public void disable(String disabledReason) {
+        log.error("Disabling periodic KRaft snapshot generation because {}.", disabledReason);
+        this.disabledReason = disabledReason;
```

Comment:
I think if we pass `null` here, it will re-enable snapshot generation. Do we want to allow that? Or should snapshot disabling be irreversible until restart?

Reply:
Yeah, I think we should not allow re-enablement. I will add a `requireNonNull`